docs/load_balancer.md

```diff
@@ -0,0 +1,278 @@
+* GCE uses the name
+* AWS can use name and/or tags
+* OpenStack uses the name
+* Rackspace can use the name and/or metadata
+
+Because the load-balancer IPs can be recreated, we are able to expose the load-balancer information through the status.
+The ServiceStatus will be extended to include publicIPs.  This will be the user-facing IP addresses or hostnames.
+
+If we require additional state in future for kube-proxy routing (for example internal load-balancer IPs),
+that will be added to a new Status field.
+
+By putting this information on Status, an external load-balancer provisioner can use the same mechanism for securely
+updating Status as kube-proxy/kubelet.
+
+
+
+## Creation flow
+
+### visibility: loadbalancer
+
+1. User `POST`s `/api/v1/namespaces/foo/services/bar` `Service{ Spec{ visibility: loadbalancer, ports: [{port: 80}] } }`
+1. API server assigns a hostPort (if the load balancer needs one), accepts and persists
+1. LB Controller wakes up, allocates a load-balancer, and `POST`s `/api/v1/namespaces/foo/services/bar` `Service{ Status{ publicIps: [ "1.2.3.4" ] }}`
+1. Service REST sets `Service.Status.publicIps = [ "1.2.3.4" ]`
+1. kube-proxy wakes up and sets iptables to receive on 1.2.3.4
+
+### visibility: public
+
+1. User `POST`s `/api/v1/namespaces/foo/services/bar` `Service{ Spec{ visibility: public, ports: [{port: 80, hostPort: 10080}] } }`
+1. API server verifies that port 10080 can be assigned, accepts and persists
+1. kube-proxy wakes up and sets iptables to receive on hostPort 10080
+
+
+In order to verify that the port can be assigned, the API server must check that the port is within a permitted range,
+and that it is not already assigned to another service.  This is similar to the allocation of internal cluster-IPs.
+The error codes & messages (in the case of port conflicts, out-of-range ports or port-exhaustion) will be analogous
+to the equivalent IP address errors.
+
+Advanced clouds (e.g. GCE) may support load-balancing without requiring a public port assignment.
+If the visibility is _loadbalancer_, a hostPort is not guaranteed to be assigned, and even if one is assigned it
+is not guaranteed to be reachable (k8s should ideally configure firewall rules to prevent load-balancer bypass).
+
+## New flags
+
+We will add a flag `public_service_ports` to the apiserver, which will determine the port-range allowed for public services.
+
+The syntax is `--public_service_ports=10000-20000` to specify the range 10000-20000 _inclusive_.
+
+(Inside the code we will likely use a different representation)
+
+The default value is `--public_service_ports=10000-20000`.  This excludes most well-known ports, and certainly excludes
+the <1024 ports.  We should try to keep the minion using ports <1024.
+
+TODO: figure out which >1024 ports are needed on the minion
+
+A null range can be specified (e.g. `--public_service_ports=0-0`).  This will have the effect of preventing any public port allocations.
```

Comment:
I just wasn't sure how this would interact with the defaults.  Happy to go with "" if the flags library allows it.  I don't think we expect people to set an empty range though, do we?